content/blog/2020-05-22-rt-template/index.md

```diff
@@ -0,0 +1,365 @@
+            }
+        }
+        /* Done entering new instructions */
+        
+        /* Remove replaced instructions from the basic block */
+        while (!Selected.empty()) {
+            Selected.front()->eraseFromParent();
+            Selected.pop_front();
+        }
+    }
+```
+The tradeoff with this approach is, this abstraction is an arbitrary one we have picked.
+The program, nor llvm provide any support. We have to create the interfaces and replacement
+with lower level tools such as basic blocks or instructions. 
+
+## Evaluation
+I evaluated this framework by implementing passes for [MachSuite][mach]. I use a mix of unoptimizes and `-O1`
+for the kernel code. I use `clang version 11.0.0` for llvm compilation and `python3.7` for Python execution. 
+I used a timing details logging to make the tests more robust. However, for certain tests this proved
+too tedious as the tests took a long time to complete.
+
+For instructions, I replaced floating point instructions with a python simulator of
+a floating point accelerator. At this flexible granularity it was fairly straight-forward to cover 
+the entire benchmark. 
+
+For functions, I created a vvadd function to test replacement for a vvadd benchmark. I also created 
+a preprocessed results reading function for functions with 2 input arguments and an output argument of 
+the same size. This function can only cover 3 microbenchmarks I wrote and 2 MachSuite benchmarks as 
+the function signature doesn't match rest of the suite. Extending the runtime library is fairly trivial,
+but incomplete.
+
+For basic blocks, I'm not maintaining functional correctness of the benchmarks. I'm replacing a basic block 
+with a floating point operator with a runtime function call. I'm collecting results to how much coverage
+this version has.
+
+For sub blocks, I tested the pass on a specific microbenchmark I wrote (`diamond`). However, this pattern
+does not appear elsewhere in MachSuite. 
+
+[mach]: https://breagen.github.io/MachSuite/
+
+## Performance
+First image illustrates performance at different abstraction levels for the `diamond` microbenchmark.
+<img src="diamond.png" width="700" >
+
+Results show instruction replacement has several orders of magnitude performance overhead.
+(C program takes micro seconds while python function calls consume minutes for a small benchmark). Function replacement 
+overhead is low, but still significant. (In the order of milliseconds). Instruction block replacement lies in
+betweent these two extremes. Runtime library in C is however more performant than the original program. I
+suspect this is due to the library compilation at a higher optimization than the original program, which is
+compiled in `-O1`.
+
+Following graph shows all the benchmarks the current function replacement functionally works. 
+
+<img src="func.png" width="700" >
+
+It is evident from this graph that for a benchmark suite as MachSuite, instruction level replacement is intractable.
```

Comment:
Well, replacing an instruction with a _Python call_ may be super slow. But it sort of depends on what you want to do.